### PR TITLE
Add glom-like searching for keys

### PR DESCRIPTION
**Pull request by UmbrellaMalware**
Implemented glom-like search in json file.
Slightly refactored the partial_read_only method
Added a class for building indexes and a class for searching in a json file

### CODE REVIEW

**Comment by UmbrellaMalware** [comment]
fix type hinting now all python version works fine

**Comment by mkrd** [comment]
@UmbrellaMalware great work! I can’t review it properly until the weekend, because I have to write my masters thesis and swore not to touch this project during the week.
But looks fantastic, I would say it only needs a few more changes:
- Rename the Searcher class to something more explicit, eg. KeySearcher
- More Tests, especially for edge cases when a subkey doesn’t exist 
- Docs also need to be updated 
- I'm also wondering how to add this to partial writing for completeness

File: dictdatabase/index_manager.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        Returns:
+          The key, start, end, indent_level, indent_with, value_hash, end
+        """
+        key_start, key_end = utils.find_outermost_key_in_json_bytes(all_file_bytes, key)
+        indent_level, indent_with = utils.detect_indentation_in_json_bytes(
+            all_file_bytes, key_start
+        )
+        value_bytes = all_file_bytes[start:end]
+        value_hash = hashlib.sha256(value_bytes).hexdigest()
+        return key, start, end, indent_level, indent_with, value_hash, end
```
Comment (mkrd) [inline dictdatabase/index_manager.py:28]: Maybe return a new class here that is a dataclass named "IndexRecord"

File: dictdatabase/io_unsafe.py [hunk 2]
```diff
@@ -79,21 +88,12 @@ def partial_read_only(db_name: str, key: str) -> dict | None:
...
-	start = key_end + (1 if all_file_bytes[key_end] == byte_codes.SPACE else 0)
-	end = utils.seek_index_through_value_bytes(all_file_bytes, start)
-
-	indent_level, indent_with  = utils.detect_indentation_in_json_bytes(all_file_bytes, key_start)
 	value_bytes = all_file_bytes[start:end]
-	value_hash = hashlib.sha256(value_bytes).hexdigest()
-
 	# Write key info to index file
-	indexer.write(key, start, end, indent_level, indent_with, value_hash, end)
+	indexer.write(*IndexManager.create_index(all_file_bytes, key, start, end))
```
Comment (mkrd) [inline dictdatabase/io_unsafe.py:96]: If we use a IndexRecord object we don’t need to spread the return value

File: dictdatabase/io_unsafe.py [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
-import json
-import hashlib
-from . import config, utils, byte_codes, indexing, io_bytes
+
+from . import byte_codes
+from . import config
+from . import indexing
+from . import io_bytes
+from . import utils
+from .index_manager import IndexManager
```
Comment (mkrd) [inline dictdatabase/io_unsafe.py:15]: I don’t think this needs to be a class, my testing showed that classes add significant overhead over function calls

File: dictdatabase/io_unsafe.py [hunk 4]
```diff
@@ -79,21 +88,12 @@ def partial_read_only(db_name: str, key: str) -> dict | None:
 
 	# Not found in index file, search for key in the entire file
 	all_file_bytes = io_bytes.read(db_name)
-	key_start, key_end = utils.find_outermost_key_in_json_bytes(all_file_bytes, key)
-
-	if key_end == -1:
+	start, end, found = Searcher().search(all_file_bytes, key)
```
Comment (mkrd) [inline dictdatabase/io_unsafe.py:91]: Also, no class required since there is no internal state. 
By the way, The overhead is about 60ns, which is double of a function call

File: dictdatabase/searching.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from typing import Tuple
+
+from dictdatabase import byte_codes
+from dictdatabase import utils
+
+
+class Searcher:
+    @staticmethod
+    def find_start_end_in_bytes(file: bytes, key: str) -> Tuple[int, int, bool]:
+        """
```
Comment (mkrd) [inline dictdatabase/searching.py:10]: Not a fan of autogenerated comments

**Comment by mkrd** [comment]
So in general, don't use classes if not required because of the overhead, except when needed. In this case, classes with __slots__ are a good compromise

File: dictdatabase/index_manager.py (hunk 1 above)
Comment (UmbrellaMalware) [reply dictdatabase/index_manager.py:28]: Yep, planned to do this in the next commits

Comment (mkrd) [reply dictdatabase/index_manager.py:28]: Okay! By the way, you can run `./cli.sh -bp` to run multiprocessing benchmarks, and see how performance is affected by using the IndexRecord

File: dictdatabase/io_unsafe.py (hunk 4 above)
Comment (UmbrellaMalware) [reply dictdatabase/io_unsafe.py:91]: Ok, but many parts are difficult to understand due to the fact that they are divided into many calls. How about replacing the Search classes and Index Manager with module files that will have all the logic with searching/building indexes?

for example:
```
key_start, key_end = utils.find_outermost_key_in_json_bytes(file, key)
if key_end == -1:
    return -1, -1, False
start = key_end + (1 if file[key_end] == byte_codes.SPACE else 0)
end = utils.seek_index_through_value_bytes(file, start)
```
        
it is difficult to understand that the beginning and end of the value of the key itself will be obtained here

`from searcher import get_value_by_key` 
in which this logic  - will be easier for understanding

**Comment by UmbrellaMalware** [comment]
The code in the project is sometimes incompatible with PEP8 how about using the Black formatter?
https://github.com/psf/black

File: dictdatabase/index_manager.py (hunk 1 above)
Comment (mkrd) [reply dictdatabase/index_manager.py:28]: Oh yeah, thats a folder that I can't share because it is out of an production project, but I was using it because it is a good example for a large file

File: dictdatabase/io_unsafe.py (hunk 4 above)
Comment (mkrd) [reply dictdatabase/io_unsafe.py:91]: Okay, that sounds good!

**Comment by mkrd** [comment]
> The code in the project is sometimes incompatible with PEP8 how about using the Black formatter? https://github.com/psf/black

Yes, when this PR is done, I will choose a formatter (I don't like black, probably gonna go for yapf)

File: tests/benchmark/run_parallel.py [hunk 6]
```diff
@@ -89,9 +89,9 @@ class Scenario:
 	ops: int = 10
 
 	def print(self):
-		res = f"✨ Scenario: {'🔹' * self.readers}{'🔻' * self.writers} ({self.readers}r{self.writers}w)"
-		res += ", 🔸 compression" if self.use_compression else ""
-		res += ", 💎 big file" if self.big_file else ""
```
Comment (mkrd) [inline tests/benchmark/run_parallel.py:94]: Was this actually an issue? I think about every terminal and editor can handle unicode by now

Comment (UmbrellaMalware) [reply tests/benchmark/run_parallel.py:94]: in my pycharm terminal this is not displayed correctly

Comment (UmbrellaMalware) [reply tests/benchmark/run_parallel.py:94]: image

**Comment by mkrd** [comment]
@UmbrellaMalware Can you re-open this PR to the branch https://github.com/mkrd/DictDataBase/tree/feat/glom-like-key-searching so I can add commits aswell? I will add you to this project as a collaborator so you can work on the branch aswell!

**Comment by UmbrellaMalware** [comment]
> @UmbrellaMalware Can you re-open this PR to the branch https://github.com/mkrd/DictDataBase/tree/feat/glom-like-key-searching so I can add commits aswell? I will add you to this project as a collaborator so you can work on the branch aswell!

Done

But, I need a few more days to finish everything I wanted for this PR, I think I will prepare a new PR over the weekend, and we can review or fix it

**Comment by mkrd** [comment]
Perfect!

File: tests/benchmark/run_parallel.py (hunk 6 above)
Comment (mkrd) [reply tests/benchmark/run_parallel.py:94]: Damn I already thought my ssd is fast, but you get more than twice the ops/sec as me!
